# Week 5 Day 4: CrewAI — Multi-Agent Collaboration, Roles & Task Delegation

Yesterday it was LangGraph, just a graph, now we are switching to CrewAI, where instead of one agent following a graph, i define a *team* of agents, each with their own role, goal, backstory plus tools as well as processes. Basically a team, a crew :D

## Goal
* Design the crew
* Create the 3 agents
* create the 3 tasks
* Run sequential process
* Then run hierarchical process
* Compare them
* Measure tokens
* Compare outputs
* Write conclusions

In [1]:
pip install -r requirements.txt

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.2.1 -> 26.1.2
[notice] To update, run: C:\Users\imama\AppData\Local\Programs\Python\Python311\python.exe -m pip install --upgrade pip


## Setup

In [7]:
pip install "crewai[groq]"

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.2.1 -> 26.1.2
[notice] To update, run: C:\Users\imama\AppData\Local\Programs\Python\Python311\python.exe -m pip install --upgrade pip


In [27]:
pip install litellm==1.74.9

  Obtaining dependency information for litellm==1.74.9 from https://files.pythonhosted.org/packages/5f/e4/f1546746049c99c6b8b247e2f34485b9eae36faa9322b84e2a17262e6712/litellm-1.74.9-py3-none-any.whl.metadata
     ---------------------------------------- 0.0/40.6 kB ? eta -:--:--
     ------------------------------ --------- 30.7/40.6 kB 1.3 MB/s eta 0:00:01
     ------------------------------ --------- 30.7/40.6 kB 1.3 MB/s eta 0:00:01
     -------------------------------------- 40.6/40.6 kB 242.7 kB/s eta 0:00:00
   ---------------------------------------- 0.0/8.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/8.7 MB 960.0 kB/s eta 0:00:10
   ---------------------------------------- 0.1/8.7 MB 812.7 kB/s eta 0:00:11
   ---------------------------------------- 0.1/8.7 MB 655.4 kB/s eta 0:00:14
    --------------------------------------- 0.2/8.7 MB 952.6 kB/s eta 0:00:09
    --------------------------------------- 0.2/8.7 MB 980.4 kB/s eta 0:00:09
   - ----------------


[notice] A new release of pip is available: 23.2.1 -> 26.1.2
[notice] To update, run: C:\Users\imama\AppData\Local\Programs\Python\Python311\python.exe -m pip install --upgrade pip


In [21]:
pip install google-genai

  Obtaining dependency information for google-genai from https://files.pythonhosted.org/packages/ba/86/5ac5fb53e44cca4a6607fb917eb331fa237c65a103b9ec2e8e8acc8a42db/google_genai-2.14.0-py3-none-any.whl.metadata
     ---------------------------------------- 0.0/55.8 kB ? eta -:--:--
     -------------------- ----------------- 30.7/55.8 kB 660.6 kB/s eta 0:00:01
     -------------------- ----------------- 30.7/55.8 kB 660.6 kB/s eta 0:00:01
     ---------------------------------- --- 51.2/55.8 kB 327.7 kB/s eta 0:00:01
     -------------------------------------- 55.8/55.8 kB 323.2 kB/s eta 0:00:00
  Obtaining dependency information for google-auth[requests]<3.0.0,>=2.56.0 from https://files.pythonhosted.org/packages/88/63/50636aae68c9bf17c891c7eb18b49baa9bd6b31d2a97b8de4813a9fc8d1c/google_auth-2.56.2-py3-none-any.whl.metadata
  Obtaining dependency information for pyasn1-modules>=0.2.1 from https://files.pythonhosted.org/packages/47/8d/d529b5d697919ba8c11ad626e835d4039be708a35b0d22de83a26


[notice] A new release of pip is available: 23.2.1 -> 26.1.2
[notice] To update, run: C:\Users\imama\AppData\Local\Programs\Python\Python311\python.exe -m pip install --upgrade pip


In [1]:
from dotenv import load_dotenv
import os, json
# from langchain_openai import ChatOpenAI

from crewai import LLM, Agent, Task, Crew, Process
from crewai.tools import tool

In [10]:
os.environ["OPENAI_API_KEY"] = os.getenv("API_KEY")
# os.environ["OPENAI_BASE_URL"] = os.getenv("BASE_URL")


In [28]:
from crewai import LLM
from dotenv import load_dotenv
import os
MODEL="gemini/gemini-3.6-flash"
load_dotenv()
llm = LLM(
    model=MODEL,
    api_key=os.getenv("API_KEY")
    # base_url=os.getenv("OPENAI_BASE_URL")
)

In [29]:
print(llm.call("Say hello"))

Hello! How can I help you today?


## Task 1: Multi-Agent Design Thinking

### Chosen Business Task:
A customer emails in asking "which laptop should I get": 
* research/compare the 3 laptops in `products.json` against their stated needs
* calculate a value score, and write them a friendly, stakeholder-ready recommendation.
* Its a mini version of a sales-engineering workflow: gather facts -> analyze/compare -> communicate


| Agent | Role | Goal | Backstory |
|---|---|---|---|
| **Product Researcher** | Data Retrieval Specialist | Pull accurate specs/prices for the laptops relevant to the customer's request, nothing more | "You are a meticulous product data analyst at an electronics retailer. You never guess a spec, you only report what's in the product catalog. You don't make recommendations — that's someone else's job." |
| **Comparison Analyst** | Quantitative Comparator | Turn raw specs into a structured, numeric comparison (value-for-money score, budget fit) | "You are a data-driven analyst who's spent years building spec-comparison spreadsheets. You care about numbers being right and defensible, not about how the message reads." |
| **Sales Communicator** | Customer-Facing Writer | Turn the analyst's comparison into a warm, non-technical recommendation email for the customer | "You are a friendly and kind customer success writer who translates spec-sheets into plain English. You've never touched the product database yourself, you just take what analysis you're handed and make it sound as human as possible" |

**Why 3 specialists could beat 1 generalist here:** each sub-step needs a different "mode": factual retrieval (precision, no embellishment), numeric reasoning (consistency), and persuasive/empathetic writing (tone) and a single agent asked to do all three in one pass tends to blend them, e.g. rounding numbers while "being friendly" or burying the actual recommendation under seats. Splitting the roles also makes each step auditable: i can check the researcher's numbers separately from the analyst's math, separately from the writer's tone.

**Where this isn't true:** for a task this tiny (3 laptops, 1 customer request), a single well-prompted agent could realistically do it in one pass with less latency and no risk of information getting lost/reformatted between agents, the overhead of 3 agents talking to each other is arguably not worth it here but let's see if we can change that.

## Task 2: Build Agents and Assign Tools

Reusing my day 1/2 tools (`calculator`) plus a new `product_lookup` tool for the catalog. I'm keeping tool access role-appropriate:

**1. Product Researcher** gets `product_lookup` only: it should be reading the catalog, not doing math or writing prose.

**2. Comparison Analyst** gets `calculator` only: it needs to compute value scores, but has no business re-querying the catalog (that's the researcher's job, and giving it lookup access would let it skip/duplicate the researcher's work).

**3. Sales Communicator** gets no tools: it's a pure writing role, it should only work off what the analyst handed it, not go fetch its own facts.

I didnt give anyone `weather_lookup` since it's irrelevant to this task, leaving it out is itself part of role-appropriate thingy


In [16]:
with open("products.json") as f:
    PRODUCTS = json.load(f)

### Tools

In [17]:
#tool 1 : product lookup
@tool("Product Lookup")
def product_lookup(product_id : str) -> str:
    """"Look up a laptop's specs by its id (e.g. laptop_a, laptop_b, laptop_c )as well as their names.
    Returns name, price, battery_life_hrs, ram_gb. Use 'all' to list every product.
    Do not hallucinate.
    """
    if product_id == "all":
        return json.dumps(PRODUCTS, indent=2)
    item = PRODUCTS.get(product_id)
    if not item:
        return f"Error: no product with id '{product_id}'. Valid ids: {list(PRODUCTS.keys())}"
    return json.dumps(item, indent=2)

#tool 2: calculator
@tool("Calculator")
def calculator(operation: str, a: float, b: float) -> str:
    """Perform arithmetic on two numbers. operation must be one of: add, subtract, multiply, divide."""
    if operation == "add":
        return str(a + b)
    elif operation == "subtract":
        return str(a - b)
    elif operation == "multiply":
        return str(a * b)
    elif operation == "divide":
        if b == 0:
            return "Error: Cannot divide by zero."
        return str(a / b)
    else:
        return "Invalid operation."

### Agents

In [18]:
researcher = Agent(
    role="Product Data Retrieval Specialist",
    goal="Pull accurate, complete specs and pricing for the laptops relevant to the customer's request, and nothing else.",
    backstory=(
        "You are a meticulous product data analyst at an electronics retailer with 8+ years of experience. "
        "You never guess a spec, you only report what's in the product catalog. "
        "You don't make recommendations or judgments about which product is 'better': that's someone else's job. "
    ),
    tools=[product_lookup],
    llm=llm,
    verbose=True,
    allow_delegation=False,
)

analyst = Agent(
    role="Quantitative Comparison Analyst",
    goal="Turn raw product specs into a structured, numeric comparison including a value-for-money score per laptop. ",
    backstory=(
        "You are a data-driven analyst who's spent several years building spec-comparison spreadsheets. "
        "You care about the numbers being right and defensible, not about how the final message reads. "
        "You always show your calculation and working, not just the result. "
    ),
    tools=[calculator],
    llm=llm,
    verbose=True,
    allow_delegation=False,
)

communicator = Agent(
    role="Customer Success Writer",
    goal="Turn the analyst's comparison into a warm, non-technical recommendation for the customer. Be nice",
    backstory=(
        "You are a customer success writer who translates spec-sheets into plain English. "
        "You've never touched the product database yourself, you just take the analysis you're handed "
        "and make it sound human, friendly, and decisive. And you are very natural with customers"
    ),
    tools=[],
    llm=llm,
    verbose=True,
    allow_delegation=False,
)


In [19]:
#just checking
analyst

Agent(role=Quantitative Comparison Analyst, goal=Turn raw product specs into a structured, numeric comparison including a value-for-money score per laptop. , backstory=You are a data-driven analyst who's spent several years building spec-comparison spreadsheets. You care about the numbers being right and defensible, not about how the final message reads. You always show your calculation and working, not just the result. )

### Tool Assignment Justification:
| Agent                  | Tools               | Why                                                                                                 |
| ---------------------- | ------------------- | --------------------------------------------------------------------------------------------------- |
| **Product Researcher** | `Product Lookup` | Needs access to the product catalog to retrieve accurate laptop information.                        |
| **Comparison Analyst** | `Calculator`    | Uses calculations to compare prices and determine value-for-money.                                  |
| **Sales Communicator** | No tools            | Only needs the outputs from the previous agents to write the final recommendation in plain English. |

## Task 3: Define Tasks & Process
Creating tasks for the agents above.

The customer's stated need (for now): *"I need a laptop mainly for work meetings and travel, battery life matters most, budget around $900."*

Each `Task` has a context list pointing at the task(s) it depends on, so CrewAI passes the earlier output(s) forward automatically instead ofmanually stitching strings together.

In [20]:
customer_request = "I need a laptop mainly for work meetings and travel, battery life matters most, budget around $900."

research_task = Task(
    description=(
        f"The customer said: '{customer_request}'. Use the Product Lookup tool with id='all' to fetch every "
        "laptop in the catalog. Report back the raw specs (name, price, battery_life_hrs, ram_gb) for ALL 3 "
        "laptops as a clean JSON-like list. Do not compare or recommend anything yet."
    ),
    expected_output=(
        "A list of exactly 3 laptops, each with name, price, battery_life_hrs, ram_gb, formatted as JSON. "
        "No commentary, no recommendation."
    ),
    agent=researcher
)

analysis_task = Task(
    description=(
        f"Customer need: '{customer_request}'. Using the specs provided by the researcher, compute a "
        "value_score for each laptop using the Calculator tool as (battery_life_hrs * 10 + ram_gb) / (price / 100), "
        "rounded to 2 decimals. Show the calculation for each laptop. Then flag which laptops fit the ~$900 budget."
    ),
    expected_output=(
        "A markdown table with columns: name, price, battery_life_hrs, ram_gb, value_score, within_budget "
        "(yes/no). One row per laptop, 3 rows total."
    ),
    agent=analyst,
    context=[research_task]
)

communication_task = Task(
    description=(
        f"Customer need: '{customer_request}'. Using the analyst's comparison table, write a short, friendly "
        "email to the customer recommending ONE laptop and briefly explaining why, in plain non-technical "
        "language. Mention 1-2 runner-up tradeoffs if relevant."
    ),
    expected_output=(
        "A 100-150 word email, with an appropriate subject, starting with 'Hi,' or any other appropriate word and ending with a clear single recommendation, no tables, "
        "no raw numbers dumped, translate them into plain language (e.g. 'great battery life' not '9.2 value_score')."
    ),
    agent=communicator,
    context=[analysis_task]
)


### Making a Sequential Crew

In [21]:
sequential_crew = Crew(
    agents = [researcher, analyst, communicator],
    tasks = [research_task, analysis_task, communication_task],
    process = Process.sequential,
    llm=llm,
    verbose=True
)

#executing crew
sequential_result = await sequential_crew.kickoff_async()
print(sequential_result)

╭─────────────────────────────────────────── 🚀 Crew Execution Started ───────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Started                                                                                         │
│  Name: crew                                                                                                     │
│  ID: f16a2280-6f0b-4fbe-8ffb-974515baa7a7                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── 📋 Task Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Started                                                                                                   │
│  Name: The customer said: 'I need a laptop mainly for work meetings and travel, battery life matters most,      │
│  budget around $900.'. Use the Product Lookup tool with id='all' to fetch every laptop in the catalog. Report   │
│  back the raw specs (name, price, battery_life_hrs, ram_gb) for ALL 3 laptops as a clean JSON-like list. Do     │
│  not compare or recommend anything yet.                                                                         │
│  ID: 7d0a9dd9-e3d1-4a6d-bb7f-5658de4bea84                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Product Data Retrieval Specialist                                                                       │
│                                                                                                                 │
│  Task: The customer said: 'I need a laptop mainly for work meetings and travel, battery life matters most,      │
│  budget around $900.'. Use the Product Lookup tool with id='all' to fetch every laptop in the catalog. Report   │
│  back the raw specs (name, price, battery_life_hrs, ram_gb) for ALL 3 laptops as a clean JSON-like list. Do     │
│  not compare or recommend anything yet.                                                                         │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

Tool product_lookup executed with result: {
  "laptop_a": {
    "name": "AeroBook 14",
    "price": 899,
    "battery_life_hrs": 10,
    "ram_gb": 16
  },
  "laptop_b": {
    "name": "SwiftPro X",
    "price": 1299,
    "battery_life_hrs": 8,...


╭──────────────────────────────────────── 🔧 Tool Execution Started (#1) ─────────────────────────────────────────╮
│                                                                                                                 │
│  Tool: product_lookup                                                                                           │
│  Args: {'product_id': 'all'}                                                                                    │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────── ✅ Tool Execution Completed (#1) ────────────────────────────────────────╮
│                                                                                                                 │
│  Tool Completed                                                                                                 │
│  Tool: product_lookup                                                                                           │
│  Output: {                                                                                                      │
│    "laptop_a": {                                                                                                │
│      "name": "AeroBook 14",                                                                                     │
│      "price": 899,                                                                                              │
│      "battery_life_hrs": 10,                                                                                    │
│      "ram_gb": 16                                                                                               │
│    },                                                                                                           │
│    "laptop_b": {                                                                                                │
│      "name": "SwiftPro X",                                                                                      │
│      "price": 1299,                                                                                             │
│      "battery_life_hrs": 8,                                                                                     │
│      "ram_gb": 32                                                                                               │
│    },                                                                                                           │
│    "laptop_c": {                                                                                                │
│      "name": "ValueBook Lite",                                                                                  │
│      "price": 549,                                                                                              │
│      "battery_life_hrs": 7,                                                                                     │
│      "ram_gb": 8                                                                                                │
│    }                                                                                                            │
│  }                                                                                                              │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[Finalize] todos_count=0, todos_with_results=0


╭───────────────────────────────────────────── ✅ Agent Final Answer ─────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Product Data Retrieval Specialist                                                                       │
│                                                                                                                 │
│  Final Answer:                                                                                                  │
│  [                                                                                                              │
│    {                                                                                                            │
│      "name": "AeroBook 14",                                                                                     │
│      "price": 899,                                                                                              │
│      "battery_life_hrs": 10,                                                                                    │
│      "ram_gb": 16                                                                                               │
│    },                                                                                                           │
│    {                                                                                                            │
│      "name": "SwiftPro X",                                                                                      │
│      "price": 1299,                                                                                             │
│      "battery_life_hrs": 8,                                                                                     │
│      "ram_gb": 32                                                                                               │
│    },                                                                                                           │
│    {                                                                                                            │
│      "name": "ValueBook Lite",                                                                                  │
│      "price": 549,                                                                                              │
│      "battery_life_hrs": 7,                                                                                     │
│      "ram_gb": 8                                                                                                │
│    }                                                                                                            │
│  ]                                                                                                              │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭────────────────────────────────────────────── 📋 Task Completion ───────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Completed                                                                                                 │
│  Name: The customer said: 'I need a laptop mainly for work meetings and travel, battery life matters most,      │
│  budget around $900.'. Use the Product Lookup tool with id='all' to fetch every laptop in the catalog. Report   │
│  back the raw specs (name, price, battery_life_hrs, ram_gb) for ALL 3 laptops as a clean JSON-like list. Do     │
│  not compare or recommend anything yet.                                                                         │
│  Agent: Product Data Retrieval Specialist                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── 📋 Task Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Started                                                                                                   │
│  Name: Customer need: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Using the specs provided by the researcher, compute a value_score for each laptop using the     │
│  Calculator tool as (battery_life_hrs * 10 + ram_gb) / (price / 100), rounded to 2 decimals. Show the           │
│  calculation for each laptop. Then flag which laptops fit the ~$900 budget.                                     │
│  ID: 09553f00-2d49-4b09-a388-96cecda1478e                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Quantitative Comparison Analyst                                                                         │
│                                                                                                                 │
│  Task: Customer need: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Using the specs provided by the researcher, compute a value_score for each laptop using the     │
│  Calculator tool as (battery_life_hrs * 10 + ram_gb) / (price / 100), rounded to 2 decimals. Show the           │
│  calculation for each laptop. Then flag which laptops fit the ~$900 budget.                                     │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────── 🔧 Tool Execution Started (#1) ─────────────────────────────────────────╮
│                                                                                                                 │
│  Tool: calculator                                                                                               │
│  Args: {'a': 116, 'b': 8.99, 'operation': 'divide'}                                                             │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

Tool calculator executed with result: 12.903225806451612...
Tool calculator executed with result: 8.622016936104696...
Tool calculator executed with result: 14.207650273224044...


╭──────────────────────────────────────── 🔧 Tool Execution Started (#2) ─────────────────────────────────────────╮
│                                                                                                                 │
│  Tool: calculator                                                                                               │
│  Args: {'b': 12.99, 'a': 112, 'operation': 'divide'}                                                            │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────── 🔧 Tool Execution Started (#3) ─────────────────────────────────────────╮
│                                                                                                                 │
│  Tool: calculator                                                                                               │
│  Args: {'b': 5.49, 'a': 78, 'operation': 'divide'}                                                              │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────── ✅ Tool Execution Completed (#3) ────────────────────────────────────────╮
│                                                                                                                 │
│  Tool Completed                                                                                                 │
│  Tool: calculator                                                                                               │
│  Output: 14.207650273224044                                                                                     │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────── ✅ Tool Execution Completed (#3) ────────────────────────────────────────╮
│                                                                                                                 │
│  Tool Completed                                                                                                 │
│  Tool: calculator                                                                                               │
│  Output: 8.622016936104696                                                                                      │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────── ✅ Tool Execution Completed (#3) ────────────────────────────────────────╮
│                                                                                                                 │
│  Tool Completed                                                                                                 │
│  Tool: calculator                                                                                               │
│  Output: 12.903225806451612                                                                                     │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[Finalize] todos_count=0, todos_with_results=0


╭───────────────────────────────────────────── ✅ Agent Final Answer ─────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Quantitative Comparison Analyst                                                                         │
│                                                                                                                 │
│  Final Answer:                                                                                                  │
│  ### Quantitative Spec Comparison & Value Analysis                                                              │
│                                                                                                                 │
│  **Formula:**                                                                                                   │
│  $$\text{value\_score} = \frac{(\text{battery\_life\_hrs} \times 10) +                                          │
│  \text{ram\_gb}}{\frac{\text{price}}{100}}$$                                                                    │
│                                                                                                                 │
│  ---                                                                                                            │
│                                                                                                                 │
│  ### Step-by-Step Calculations                                                                                  │
│                                                                                                                 │
│  1. **AeroBook 14**                                                                                             │
│     * **Numerator:** $(10 \text{ hrs} \times 10) + 16 \text{ GB} = 100 + 16 = 116$                              │
│     * **Denominator:** $\$899 / 100 = 8.99$                                                                     │
│     * **Value Score:** $116 / 8.99 = 12.9032... \approx \mathbf{12.90}$                                         │
│     * **Budget Check:** $\$899 \le \$900 \rightarrow \mathbf{yes}$                                              │
│                                                                                                                 │
│  2. **SwiftPro X**                                                                                              │
│     * **Numerator:** $(8 \text{ hrs} \times 10) + 32 \text{ GB} = 80 + 32 = 112$                                │
│     * **Denominator:** $\$1299 / 100 = 12.99$                                                                   │
│     * **Value Score:** $112 / 12.99 = 8.6220... \approx \mathbf{8.62}$                                          │
│     * **Budget Check:** $\$1299 > \$900 \rightarrow \mathbf{no}$                                                │
│                                                                                                                 │
│  3. **ValueBook Lite**                                                                                          │
│     * **Numerator:** $(7 \text{ hrs} \times 10) + 8 \text{ GB} = 70 + 8 = 78$                                   │
│     * **Denominator:** $\$549 / 100 = 5.49$                                                                     │
│     * **Value Score:** $78 / 5.49 = 14.2076... \approx \mathbf{14.21}$                                          │
│     * **Budget Check:** $\$549 \le \$900 \rightarrow \mathbf{yes}$                                              │
│                                                                                                                 │
│  ---                                                                                                            │
│                                                        

╭────────────────────────────────────────────── 📋 Task Completion ───────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Completed                                                                                                 │
│  Name: Customer need: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Using the specs provided by the researcher, compute a value_score for each laptop using the     │
│  Calculator tool as (battery_life_hrs * 10 + ram_gb) / (price / 100), rounded to 2 decimals. Show the           │
│  calculation for each laptop. Then flag which laptops fit the ~$900 budget.                                     │
│  Agent: Quantitative Comparison Analyst                                                                         │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── 📋 Task Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Started                                                                                                   │
│  Name: Customer need: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Using the analyst's comparison table, write a short, friendly email to the customer             │
│  recommending ONE laptop and briefly explaining why, in plain non-technical language. Mention 1-2 runner-up     │
│  tradeoffs if relevant.                                                                                         │
│  ID: 3e046650-2920-4d95-9eb9-fa65d4972382                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Customer Success Writer                                                                                 │
│                                                                                                                 │
│  Task: Customer need: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Using the analyst's comparison table, write a short, friendly email to the customer             │
│  recommending ONE laptop and briefly explaining why, in plain non-technical language. Mention 1-2 runner-up     │
│  tradeoffs if relevant.                                                                                         │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

[Finalize] todos_count=0, todos_with_results=0


╭───────────────────────────────────────────── ✅ Agent Final Answer ─────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Customer Success Writer                                                                                 │
│                                                                                                                 │
│  Final Answer:                                                                                                  │
│  Subject: The perfect laptop for your travel and meetings!                                                      │
│                                                                                                                 │
│  Hi there,                                                                                                      │
│                                                                                                                 │
│  I took a close look at options for your travel schedule and long meeting days, and I have the perfect match!   │
│                                                                                                                 │
│  I strongly recommend the **AeroBook 14**. Sitting right at your $900 budget, it gives you phenomenal all-day   │
│  battery life for long flights and back-to-back calls, plus plenty of speed to keep all your work apps running  │
│  smoothly.                                                                                                      │
│                                                                                                                 │
│  As for trade-offs, you could save money with the ValueBook Lite, but you’d give up valuable battery time and   │
│  performance. The SwiftPro X offers extra power, but it stretches way past your budget and surprisingly offers  │
│  less battery life.                                                                                             │
│                                                                                                                 │
│  Overall, the **AeroBook 14** is definitely your best companion for staying productive on the go. Let me know   │
│  if you'd like help getting set up!                                                                             │
│                                                                                                                 │
│  Best,                                                                                                          │
│  Customer Success Team                                                                                          │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭────────────────────────────────────────────── 📋 Task Completion ───────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Completed                                                                                                 │
│  Name: Customer need: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Using the analyst's comparison table, write a short, friendly email to the customer             │
│  recommending ONE laptop and briefly explaining why, in plain non-technical language. Mention 1-2 runner-up     │
│  tradeoffs if relevant.                                                                                         │
│  Agent: Customer Success Writer                                                                                 │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── Crew Completion ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Completed                                                                                       │
│  Name: crew                                                                                                     │
│  ID: f16a2280-6f0b-4fbe-8ffb-974515baa7a7                                                                       │
│  Final Output: Subject: The perfect laptop for your travel and meetings!                                        │
│                                                                                                                 │
│  Hi there,                                                                                                      │
│                                                                                                                 │
│  I took a close look at options for your travel schedule and long meeting days, and I have the perfect match!   │
│                                                                                                                 │
│  I strongly recommend the **AeroBook 14**. Sitting right at your $900 budget, it gives you phenomenal all-day   │
│  battery life for long flights and back-to-back calls, plus plenty of speed to keep all your work apps running  │
│  smoothly.                                                                                                      │
│                                                                                                                 │
│  As for trade-offs, you could save money with the ValueBook Lite, but you’d give up valuable battery time and   │
│  performance. The SwiftPro X offers extra power, but it stretches way past your budget and surprisingly offers  │
│  less battery life.                                                                                             │
│                                                                                                                 │
│  Overall, the **AeroBook 14** is definitely your best companion for staying productive on the go. Let me know   │
│  if you'd like help getting set up!                                                                             │
│                                                                                                                 │
│  Best,                                                                                                          │
│  Customer Success Team                                                                                          │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

Subject: The perfect laptop for your travel and meetings!

Hi there,

I took a close look at options for your travel schedule and long meeting days, and I have the perfect match!

I strongly recommend the **AeroBook 14**. Sitting right at your $900 budget, it gives you phenomenal all-day battery life for long flights and back-to-back calls, plus plenty of speed to keep all your work apps running smoothly.

As for trade-offs, you could save money with the ValueBook Lite, but you’d give up valuable battery time and performance. The SwiftPro X offers extra power, but it stretches way past your budget and surprisingly offers less battery life.

Overall, the **AeroBook 14** is definitely your best companion for staying productive on the go. Let me know if you'd like help getting set up!

Best,  
Customer Success Team


╭──────────────────────────────────────────────── Tracing Status ─────────────────────────────────────────────────╮
│                                                                                                                 │
│  Info: Tracing is disabled.                                                                                     │
│                                                                                                                 │
│  To enable tracing, do any one of these:                                                                        │
│  • Set tracing=True in your Crew/Flow code                                                                      │
│  • Set CREWAI_TRACING_ENABLED=true in your project's .env file                                                  │
│  • Run: crewai traces enable                                                                                    │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

### Note one place where a task's output wasn't in the format the next agent needed — how did you fix the prompt/expected_output to solve it?
One issue was that the Product Researcher initially returned laptop details as plain text, which made it difficult for the Comparison Analyst to extract the values needed for comparison. I fixed this by updating the research task's expected_output to require a structured list of each laptop's name, price, battery life, and RAM, ensuring the analyst received the data in a consistent format.

## Task 4: Hierarchical Delegation

Same 3 workers, but now a manager agent sits on top, breaks down the goal, delegates to whichever worker it thinks is appropriate, and reviews the output before finishing. In `Process.hierarchical` i don't pass explicit Task->Task context chains, the manager is responsible for sequencing that.

In [22]:
# manager extension
manager = Agent(
    role="Sales Ops Manager",
    goal="Deliver a correct, well-written laptop recommendation to the customer by delegating research, analysis, and writing to the right specialist and reviewing their work.",
    backstory=(
        "You are a sales operations manager who doesn't do the hands-on work yourself. "
        "You break the goal into sub-tasks, assign each to the right specialist on your team, "
        "and check their output before passing it along or accepting the final result."
    ),
    llm=llm,
    verbose=True,
    allow_delegation=True,
)




In [26]:
## manager's task
overall_task = Task(
    description=(
        f"Customer said: '{customer_request}'. Produce a final customer-facing email recommending exactly one "
        "laptop from the catalog, backed by a correct value-score comparison of all 3 laptops. "
        "Delegate the catalog lookup to the researcher, the numeric comparison to the analyst, "
        "and the final email to the communicator."
    ),
    expected_output=(
        "A 100-150 word customer email recommending one laptop, with the reasoning grounded in a correct "
        "comparison of all 3 laptops (battery life, price, RAM, budget fit)."
    ),
    # agent=manager,
)


In [31]:

hierarchical_crew = Crew(
    agents=[researcher, analyst, communicator],
    tasks=[overall_task],
    process=Process.hierarchical,
    manager_agent=manager,
    verbose=True
)

hierarchical_result = await hierarchical_crew.kickoff_async()
print(hierarchical_result)

╭─────────────────────────────────────────── 🚀 Crew Execution Started ───────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Started                                                                                         │
│  Name: crew                                                                                                     │
│  ID: ca6a7dec-33cf-4b9c-9be8-d80764f2edc5                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── 📋 Task Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Started                                                                                                   │
│  Name: Customer said: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Produce a final customer-facing email recommending exactly one laptop from the catalog, backed  │
│  by a correct value-score comparison of all 3 laptops. Delegate the catalog lookup to the researcher, the       │
│  numeric comparison to the analyst, and the final email to the communicator.                                    │
│  ID: 2279c639-f0bb-4ef4-a6d3-5165c8f263eb                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Sales Ops Manager                                                                                       │
│                                                                                                                 │
│  Task: Customer said: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Produce a final customer-facing email recommending exactly one laptop from the catalog, backed  │
│  by a correct value-score comparison of all 3 laptops. Delegate the catalog lookup to the researcher, the       │
│  numeric comparison to the analyst, and the final email to the communicator.                                    │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

ERROR:root:Google Gemini API error: 429 - You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To monitor your current usage, head to: https://ai.dev/rate-limit. 
* Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit: 20, model: gemini-3.6-flash
Please retry in 21.979252598s.


An unknown error occurred. Please check the details below.
Error details: 429 RESOURCE_EXHAUSTED. {'error': {'code': 429, 'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To monitor your current usage, head to: https://ai.dev/rate-limit. \n* Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit: 20, model: gemini-3.6-flash\nPlease retry in 21.979252598s.', 'status': 'RESOURCE_EXHAUSTED', 'details': [{'@type': 'type.googleapis.com/google.rpc.Help', 'links': [{'description': 'Learn more about Gemini API quotas', 'url': 'https://ai.google.dev/gemini-api/docs/rate-limits'}]}, {'@type': 'type.googleapis.com/google.rpc.QuotaFailure', 'violations': [{'quotaMetric': 'generativelanguage.googleapis.com/generate_content_free_tier_requests', 'quotaId': 'GenerateRequestsPerDayPerProjectPerModel-FreeTier', 'quotaDime

ERROR:crewai.flow.runtime:Error executing listener call_llm_native_tools: 429 RESOURCE_EXHAUSTED. {'error': {'code': 429, 'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To monitor your current usage, head to: https://ai.dev/rate-limit. \n* Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit: 20, model: gemini-3.6-flash\nPlease retry in 21.979252598s.', 'status': 'RESOURCE_EXHAUSTED', 'details': [{'@type': 'type.googleapis.com/google.rpc.Help', 'links': [{'description': 'Learn more about Gemini API quotas', 'url': 'https://ai.google.dev/gemini-api/docs/rate-limits'}]}, {'@type': 'type.googleapis.com/google.rpc.QuotaFailure', 'violations': [{'quotaMetric': 'generativelanguage.googleapis.com/generate_content_free_tier_requests', 'quotaId': 'GenerateRequestsPerDayPerProjectPerModel-FreeTier', 'quotaDime

╭───────────────────────────────────────────────── ❌ LLM Error ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  LLM Call Failed                                                                                                │
│  Error: Google Gemini API error: 429 - You exceeded your current quota, please check your plan and billing      │
│  details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To    │
│  monitor your current usage, head to: https://ai.dev/rate-limit.                                                │
│  * Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit:     │
│  20, model: gemini-3.6-flash                                                                                    │
│  Please retry in 21.979252598s.                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

An unknown error occurred. Please check the details below.
Error details: 429 RESOURCE_EXHAUSTED. {'error': {'code': 429, 'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To monitor your current usage, head to: https://ai.dev/rate-limit. \n* Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit: 20, model: gemini-3.6-flash\nPlease retry in 21.979252598s.', 'status': 'RESOURCE_EXHAUSTED', 'details': [{'@type': 'type.googleapis.com/google.rpc.Help', 'links': [{'description': 'Learn more about Gemini API quotas', 'url': 'https://ai.google.dev/gemini-api/docs/rate-limits'}]}, {'@type': 'type.googleapis.com/google.rpc.QuotaFailure', 'violations': [{'quotaMetric': 'generativelanguage.googleapis.com/generate_content_free_tier_requests', 'quotaId': 'GenerateRequestsPerDayPerProjectPerModel-FreeTier', 'quotaDime

╭──────────────────────────────────────────────── 📋 Task Failure ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Failed                                                                                                    │
│  Name: Customer said: 'I need a laptop mainly for work meetings and travel, battery life matters most, budget   │
│  around $900.'. Produce a final customer-facing email recommending exactly one laptop from the catalog, backed  │
│  by a correct value-score comparison of all 3 laptops. Delegate the catalog lookup to the researcher, the       │
│  numeric comparison to the analyst, and the final email to the communicator.                                    │
│  Agent: Sales Ops Manager                                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭───────────────────────────────────────────────── Crew Failure ──────────────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Failed                                                                                          │
│  Name: crew                                                                                                     │
│  ID: ca6a7dec-33cf-4b9c-9be8-d80764f2edc5                                                                       │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── Tracing Status ─────────────────────────────────────────────────╮
│                                                                                                                 │
│  Info: Tracing is disabled.                                                                                     │
│                                                                                                                 │
│  To enable tracing, do any one of these:                                                                        │
│  • Set tracing=True in your Crew/Flow code                                                                      │
│  • Set CREWAI_TRACING_ENABLED=true in your project's .env file                                                  │
│  • Run: crewai traces enable                                                                                    │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

ClientError: 429 RESOURCE_EXHAUSTED. {'error': {'code': 429, 'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To monitor your current usage, head to: https://ai.dev/rate-limit. \n* Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit: 20, model: gemini-3.6-flash\nPlease retry in 21.979252598s.', 'status': 'RESOURCE_EXHAUSTED', 'details': [{'@type': 'type.googleapis.com/google.rpc.Help', 'links': [{'description': 'Learn more about Gemini API quotas', 'url': 'https://ai.google.dev/gemini-api/docs/rate-limits'}]}, {'@type': 'type.googleapis.com/google.rpc.QuotaFailure', 'violations': [{'quotaMetric': 'generativelanguage.googleapis.com/generate_content_free_tier_requests', 'quotaId': 'GenerateRequestsPerDayPerProjectPerModel-FreeTier', 'quotaDimensions': {'location': 'global', 'model': 'gemini-3.6-flash'}, 'quotaValue': '20'}]}, {'@type': 'type.googleapis.com/google.rpc.RetryInfo', 'retryDelay': '21s'}]}}

### Sequential vs Hierarchical

| | Pros | Cons | When to use |
|---|---|---|---|
| **Sequential** | Predictable order, cheapest (no manager overhead), easy to debug | I have to pre-decide the pipeline; can't adapt if one step needs to loop back or be redone; brittle if a task genuinely needs re-ordering per input | Task shape is well understood and stable, like this one — fixed research -> analyze -> write pipeline |
| **Hierarchical** | Manager can re-delegate, ask for redo, or reorder if something's missing; adapts better to messier/more open-ended goals | More LLM calls (manager reasoning + delegation overhead) = more tokens/latency/cost, and an extra point where things can go wrong (manager mis-delegating) | Task is more open-ended, sub-steps aren't fixed in advance, or you want a review/QA layer on top of workers |

For this specific task the manager's plan matched the sequential order almost exactly — makes sense, since the dependency chain (research -> analyze -> write) is basically the only sensible order. The main practical difference i actually noticed was extra manager "thinking" turns in the log before each delegation, which is exactly where the added cost comes from (see Task 5).


## Task 5: Evaluation & Cost Awareness

CrewAI exposes token usage on the crew's output via `.token_usage`, which i'm using

In [32]:
def summarize_usage(result, label):
    usage = result.token_usage
    print(f"--- {label} ---")
    print(f"prompt_tokens:     {usage.prompt_tokens}")
    print(f"completion_tokens: {usage.completion_tokens}")
    print(f"total_tokens:      {usage.total_tokens}")
    print(f"successful_requests: {usage.successful_requests}")
    return usage

seq_usage = summarize_usage(sequential_result, "Sequential Crew")
hier_usage = summarize_usage(hierarchical_result, "Hierarchical Crew")


--- Sequential Crew ---
prompt_tokens:     10554
completion_tokens: 11745
total_tokens:      22299
successful_requests: 18
--- Hierarchical Crew ---
prompt_tokens:     27660
completion_tokens: 37396
total_tokens:      65056
successful_requests: 36


In [34]:
# using a placeholder rate here since my endpoint is a custom gateway, not a metered public API.
RATE_PER_1K_TOKENS = 0.002

def estimate_cost(usage):
    return round((usage.total_tokens / 1000) * RATE_PER_1K_TOKENS, 5)

print("Sequential estimated cost:   $", estimate_cost(seq_usage))
print("Hierarchical estimated cost: $", estimate_cost(hier_usage))
print("Day 3 Single-agent langgraph estimated cost: $0.015–$0.020")


Sequential estimated cost:   $ 0.0446
Hierarchical estimated cost: $ 0.13011
Day 3 Single-agent langgraph estimated cost: $0.015–$0.020


### Token Usage Comparison

| Approach | Prompt Tokens | Completion Tokens | Total Tokens |
|----------|--------------:|------------------:|-------------:|
| **Single-Agent (Day 3)\*** | 8,000 | 1,500 | 9,500 |
| **Sequential Crew** | 10,554 | 11,745 | 22,299 |
| **Hierarchical Crew** | 27,660 | 37,396 | 65,056 |

### Cost & Performance Comparison

| Approach | Approx. Cost | Token Usage | Cost Level | Notes |
|----------|-------------:|-------------|------------|-------|
| **Single-Agent (Day 3)\*** | **\$0.0190** | Low | Low | Single model handled the entire workflow efficiently. |
| **Sequential Crew** | **\$0.0446** | Moderate | Moderate | Three specialist agents worked sequentially and produced accurate, well-structured output. |
| **Hierarchical Crew** | **\$0.1301** | High | High | Manager delegation increased token usage significantly. Delegation issues also reduced output quality. |

> **\*** *Day 3 values are approximate because the original token logs were unavailable.*


### Success Criteria and Manual Evaluation
**Success Criteria**
1. **Factual grounding** : does the recommendation email correctly reflect the actual catalog data (right price, right battery life, no invented specs)?
2. **Completeness** :does it address the customer's stated priorities (battery life, budget) and mention at least one runner-up tradeoff?
3. **Tone** : does it read as a natural, friendly customer email rather than a spec dump?

| Run                  | Factual Grounding | Completeness | Customer-Friendly Tone | Overall   |
| -------------------- | ----------------- | ------------ | ---------------------- | --------- |
| Single-Agent (Day 3) | **5/5**           | **4/5**      | **4/5**                | **13/15** |
| Sequential Crew      | **5/5**           | **5/5**      | **5/5**                | **15/15** |
| Hierarchical Crew    | **2/5**           | **3/5**      | **5/5**                | **10/15** |




### **Reflection:** 

For this task, the sequential multi-agent crew was worthwhile because each agent had a clear responsibility, resulting in an accurate and well-structured recommendation. 

However, the hierarchical approach introduced additional complexity and token usage while also encountering delegation issues, making it less reliable in practice. Compared to the single-agent LangGraph solution, the crew improved task organization but required significantly more resources. For a relatively small workflow like this, a well-designed single agent or sequential crew is more practical than a hierarchical setup. Sequential was the most balanced one out of all
